tests/ttnn/unit_tests/operations/test_moreh_dot.py

```diff
@@ -87,17 +103,93 @@ def get_tensors(
         [1, 1, 1, 323],  # test multiple tiles, not a multiple of 32
     ),
 )
-def test_moreh_matmul_1d(input_shape, device):
+@pytest.mark.parametrize(
+    "dtype",
+    (
+        ttnn.float32,
+        ttnn.bfloat16,
+        ttnn.int32,
+        None,
+    ),
+)
+def test_moreh_matmul_1d(input_shape, dtype, device):
     torch.manual_seed(3072)
     # get tensors
     output_shape = [1, 1, 1, 1]
+    if dtype is None:
+        npu_dtype = torch.bfloat16
+    npu_dtype = dtype
+    cpu_dtype = get_torch_type(dtype)
+
     tt_input, tt_other, _, _, _, _, torch_input, torch_other, _ = get_tensors(
-        input_shape, input_shape, output_shape, False, False, True, device
+        input_shape, input_shape, output_shape, False, False, True, device, npu_dtype, cpu_dtype
     )
 
     # tt matmul
     cpu_layout = ttnn.ROW_MAJOR_LAYOUT
-    tt_out = ttnn.moreh_dot(tt_input, tt_other).cpu().to(cpu_layout).unpad_from_tile(output_shape).to_torch()
+    tt_out = (
+        ttnn.operations.moreh.dot(tt_input, tt_other, dtype=dtype)
+        .cpu()
+        .to(cpu_layout)
+        .unpad_from_tile(output_shape)
+        .to_torch()
+    )
+
+    # torch matmul
+    torch_input = torch.reshape(torch_input, (torch_input.shape[-1],))
+    torch_other = torch.reshape(torch_other, (torch_other.shape[-1],))
+    torch_out = torch.matmul(torch_input, torch_other)
+
+    # test for equivalance
+    rtol = atol = 0.1
+    passing, output_pcc = comp_allclose_and_pcc(torch_out, tt_out[0][0][0][0], pcc=0.999, rtol=rtol, atol=atol)
+    logger.debug(f"Out passing={passing}")
+    logger.debug(f"Output pcc={output_pcc}")
+
+    assert passing
+
+
+@pytest.mark.parametrize(
+    "input_shape",
+    (
+        [1, 1, 1, 10],  # test not mutiple of 32 case
+        [1, 1, 1, 32],  # test single tile
+        [1, 1, 1, 352],  # test multiple tiles
+        [1, 1, 1, 323],  # test multiple tiles, not a multiple of 32
+    ),
+)
+@pytest.mark.parametrize(
+    "dtype",
+    (
+        ttnn.float32,
+        ttnn.bfloat16,
+        ttnn.int32,
+        None,
+    ),
+)
+def test_moreh_matmul_1d_callback(input_shape, dtype, device, use_program_cache):
+    torch.manual_seed(3072)
+    # get tensors
+    output_shape = [1, 1, 1, 1]
+    if dtype is None:
+        npu_dtype = torch.bfloat16
+    npu_dtype = dtype
+    cpu_dtype = get_torch_type(dtype)
+
+    tt_input, tt_other, _, _, _, _, torch_input, torch_other, _ = get_tensors(
+        input_shape, input_shape, output_shape, False, False, True, device, npu_dtype, cpu_dtype
+    )
+
+    # tt matmul
+    cpu_layout = ttnn.ROW_MAJOR_LAYOUT
+    for i in range(2):
+        tt_out = (
+            ttnn.operations.moreh.dot(tt_input, tt_other, dtype=dtype)
+            .cpu()
+            .to(cpu_layout)
+            .unpad_from_tile(output_shape)
+            .to_torch()
+        )
```

Comment:
In the current approach, the same input and output tensors are reused in the second loop, which could result in the test passing without updating the tensor address in the callback function.

It would be more effective to create the tensors entirely within the loop.
Reference: https://github.com/tenstorrent/tt-metal/pull/13139